File: Data Structures & Algorithms/number-of-visible-people-in-a-queue/submission-1.cpp
class Solution {
public:
    vector<int> canSeePersonsCount(vector<int>& heights) {
        int n = heights.size();
        vector<int> res(n, 0);
        stack<int> st;

        for (int i = n - 1; i >= 0; --i) {
            while (!st.empty() && st.top() < heights[i]) {
                st.pop();
                res[i]++;
            }
            if (!st.empty()) {
                res[i]++;
            }
            st.push(heights[i]);
        }

        return res;
    }
};

File: Data Structures & Algorithms/daily-temperatures/submission-1.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<int[]> st = new Stack<>();
        int[] res = new int[temperatures.length];
        //(30,0) (38,1)
        for (int i = 0; i < temperatures.length; i++) {
            while (!st.isEmpty() && st.peek()[0] < temperatures[i]) {
                res[st.peek()[1]] = i - st.peek()[1];
                st.pop();
            }
            st.add(new int[]{temperatures[i], i});
        }
        return res;
    }
}
